pallets/dapp-staking-v3/src/lib.rs

```diff
@@ -567,6 +713,219 @@ pub mod pallet {
 
             Ok(())
         }
+
+        /// Stake the specified amount on a smart contract.
+        /// The `amount` specified **must** be available for staking and meet the required minimum, otherwise the call will fail.
+        ///
+        /// Depending on the period type, appropriate stake amount will be updated.
+        #[pallet::call_index(9)]
+        #[pallet::weight(Weight::zero())]
+        pub fn stake(
+            origin: OriginFor<T>,
+            smart_contract: T::SmartContract,
+            #[pallet::compact] amount: Balance,
+        ) -> DispatchResult {
+            Self::ensure_pallet_enabled()?;
+            let account = ensure_signed(origin)?;
+
+            ensure!(amount > 0, Error::<T>::ZeroAmount);
+
+            ensure!(
+                Self::is_active(&smart_contract),
+                Error::<T>::NotOperatedDApp
+            );
+
+            let protocol_state = ActiveProtocolState::<T>::get();
+            // Staker always stakes from the NEXT era
+            let stake_era = protocol_state.era.saturating_add(1);
+            ensure!(
+                !protocol_state.period_info.is_next_period(stake_era),
+                Error::<T>::PeriodEndsInNextEra
+            );
+
+            let mut ledger = Ledger::<T>::get(&account);
+
+            // 1.
+            // Increase stake amount for the next era & current period in staker's ledger
+            ledger
+                .add_stake_amount(amount, stake_era, protocol_state.period_number())
+                .map_err(|err| match err {
+                    AccountLedgerError::InvalidPeriod => {
+                        Error::<T>::UnclaimedRewardsFromPastPeriods
+                    }
+                    AccountLedgerError::UnavailableStakeFunds => Error::<T>::UnavailableStakeFunds,
+                    AccountLedgerError::NoCapacity => Error::<T>::TooManyStakeChunks,
+                    // Defensive check, should never happen
+                    _ => Error::<T>::InternalStakeError,
+                })?;
```

Comment:
If we expect to have conversion b/w`AccountLedgerError` and DispatchError, then IMO we should implement `From` trait over the type.

Reply:
I did think of that, but for different `AccountLedgerError` values, we get different pallet errors, depending on the context (`stake` or `unstake` functions).